File: twitterScraper/twitterScraper.py
import twitter

"""
Downloads all tweets from a given user.
Uses twitter.Api.GetUserTimeline to retreive the last 3,200 tweets from a user.
Twitter doesn't allow retreiving more tweets than this through the API, so we get
as many as possible.
t.py should contain the imported variables.
"""

import json
import sys
import csv
import os

import twitter

CONSUMER_KEY = 'vlCeSQPk3vbxgHOLDRHW4HyEF'
CONSUMER_SECRET = 'sHeBkkGgD05hFNHK2IdcbE17BXXjah6CvAkI21WAq6ta6DDP96'
ACCESS_TOKEN_KEY = '1273577244-6wUVJKr8LEbSy7ZdeRbNF9GwVfYcrerAzvktJYw'
ACCESS_TOKEN_SECRET = 'XUX5UJpyyUZoQCpBulPjAgq7PAJbIWheFkjTlEN8zzGcL'


def get_tweets(api=None, screen_name=None, count=None):
    timeline = api.GetUserTimeline(screen_name=screen_name, count=count)
    for tweet in timeline:
        try:
            created_at = tweet.created_at
            username = screen_name
            text = " ".join(tweet.full_text.split("\n"))
            likes = tweet.favorite_count
            retweets = tweet.retweet_count
            hashtags = tweet.hashtags

            write_csv([created_at, username, text, likes, retweets, hashtags])
        except Exception as err:
            print(err)
            pass


def write_csv(list):
    if not os.path.isfile(f'tweets/tweets_{list[1]}.csv'):
        with open(f'tweets/tweets_{list[1]}.csv', mode='w') as csv_file:
            fieldnames = ['created_at', 'username', 'text', 'likes', 'retweets', 'hashtags']
            writer = csv.DictWriter(csv_file, fieldnames=fieldnames)
            writer.writeheader()

    with open(f'tweets/tweets_{list[1]}.csv', mode='a+') as csv_file:
        fieldnames = ['created_at', 'username', 'text', 'likes', 'retweets', 'hashtags']
        writer = csv.DictWriter(csv_file, fieldnames=fieldnames)

        writer.writerow({'created_at': list[0],
                         'username': list[1],
                         'text': list[2],
                         'likes': list[3],
                         'retweets': list[4],
                         'hashtags': list[5]})

if __name__ == "__main__":
    api = twitter.Api(
        CONSUMER_KEY, CONSUMER_SECRET, ACCESS_TOKEN_KEY, ACCESS_TOKEN_SECRET, tweet_mode='extended'
    )
    screen_name = sys.argv[1]
    count = sys.argv[2]
    timeline = get_tweets(api=api, screen_name=screen_name, count=count)

File: twitterScraperGo/twitterScraper.go
package main

// Reads as much of a user's last 3200 public Tweets as the Twitter API
// returns, and prints each Tweet to a file.  This example functions
// the same as user_timeline, but uses application-only auth.

import (
	"encoding/json"
	"flag"
	"fmt"
	"github.com/kurrik/oauth1a"
	"github.com/kurrik/twittergo"
	"io/ioutil"
	"net/http"
	"net/url"
	"os"
	"strconv"
	"strings"
	"time"
	"encoding/csv"
)

var totalTweets []CsvTweet
var NAME = os.Args[1]

func LoadCredentials() (client *twittergo.Client, err error) {
	credentials, err := ioutil.ReadFile("CREDENTIALS")
	if err != nil {
		return
	}
	lines := strings.Split(string(credentials), "\n")
	config := &oauth1a.ClientConfig{
		ConsumerKey:    lines[0],
		ConsumerSecret: lines[1],
	}
	client = twittergo.NewClient(config, nil)
	return
}

type Args struct {
	ScreenName string
	OutputFile string
}

func parseArgs() *Args {
	a := &Args{}
	flag.StringVar(&a.ScreenName, "screen_name", NAME, "Screen name")
	flag.StringVar(&a.OutputFile, "out", fmt.Sprintf("./Json/%v.json", NAME), "Output file")
	flag.Parse()
	return a
}

type CsvTweet struct {
	Fecha 		string 	`json:"created_at"`
	Usuario 	string 	`json:"name"`
	Texto 		string 	`json:"full_text"`
	Likes 		int 	`json:"favorite_count"`
	Retweets 	int 	`json:"retweet_count"`
}

func writeCsv(args *Args) error{
	filePath := fmt.Sprintf("./Tweets/tweets_%v.csv", NAME)
	csvFile, err := os.Create(filePath)
	fmt.Println(args.ScreenName)

	if err != nil {
		fmt.Println(err)
		return err
	}
	defer csvFile.Close()

	writer := csv.NewWriter(csvFile)

	for _, tweets := range totalTweets {
		var record []string
		record = append(record, tweets.Fecha)
		record = append(record, NAME)
		record = append(record, strings.Join(strings.Split(tweets.Texto, "\n"), " "))
		record = append(record, strconv.Itoa(tweets.Likes))
		record = append(record, strconv.Itoa(tweets.Retweets))
		writer.Write(record)
	}

	// remember to flush!
	writer.Flush()
	return nil
}

func parseJson(tweet []byte) error{
	var readJson CsvTweet
	err := json.Unmarshal([]byte(tweet), &readJson)
	if err != nil {
		fmt.Println(err)
		return err
	}
	totalTweets = append(totalTweets, readJson)
	return nil
}

func main() {
	var (
		err     error
		client  *twittergo.Client
		req     *http.Request
		resp    *twittergo.APIResponse
		args    *Args
		max_id  uint64
		out     *os.File
		query   url.Values
		results *twittergo.Timeline
		text    []byte
	)
	args = parseArgs()
	if client, err = LoadCredentials(); err != nil {
		fmt.Printf("Could not parse CREDENTIALS file: %v\n", err)
		os.Exit(1)
	}
	if out, err = os.Create(args.OutputFile); err != nil {
		fmt.Printf("Could not create output file: %v\n", args.OutputFile)
		os.Exit(1)
	}
	defer out.Close()
	const (
		count   int = 200
		urltmpl     = "/1.1/statuses/user_timeline.json?%v"
		minwait     = time.Duration(10) * time.Second
	)
	query = url.Values{}
	query.Set("exclude_replies", "true")
	query.Set("include_rts", "false")
	query.Set("tweet_mode", "extended")
	query.Set("count", fmt.Sprintf("%v", count))
	query.Set("screen_name", args.ScreenName)
	total := 0
	for {
		if max_id != 0 {
			query.Set("max_id", fmt.Sprintf("%v", max_id))
		}
		endpoint := fmt.Sprintf(urltmpl, query.Encode())
		if req, err = http.NewRequest("GET", endpoint, nil); err != nil {
			fmt.Printf("Could not parse request: %v\n", err)
			os.Exit(1)
		}
		if resp, err = client.SendRequest(req); err != nil {
			fmt.Printf("Could not send request: %v\n", err)
			os.Exit(1)
		}
		results = &twittergo.Timeline{}
		if err = resp.Parse(results); err != nil {
			if rle, ok := err.(twittergo.RateLimitError); ok {
				dur := rle.Reset.Sub(time.Now()) + time.Second
				if dur < minwait {
					// Don't wait less than minwait.
					dur = minwait
				}
				msg := "Rate limited. Reset at %v. Waiting for %v\n"
				fmt.Printf(msg, rle.Reset, dur)
				time.Sleep(dur)
				continue // Retry request.
			} else {
				fmt.Printf("Problem parsing response: %v\n", err)
			}
		}
		batch := len(*results)
		if batch == 0 {
			fmt.Printf("No more results, end of timeline.\n")
			break
		}
		for _, tweet := range *results {
			if text, err = json.Marshal(tweet); err != nil {
				fmt.Printf("Could not encode Tweet: %v\n", err)
				os.Exit(1)
			}
			err = parseJson(text)
			if err != nil {
				fmt.Println("Hubo un error al parsear el JSON")
			}
			out.Write(text)
			out.Write([]byte("\n"))
			max_id = tweet.Id() - 1
			total += 1
		}
		err = writeCsv(args)
		if err != nil {
			fmt.Println("Hubo un error en la escritura")
		}

		fmt.Printf("Got %v Tweets from %v", batch, NAME)
		if resp.HasRateLimit() {
			fmt.Printf(", %v calls available", resp.RateLimitRemaining())
		}
		fmt.Printf(".\n")

	}
	fmt.Printf("--------------------------------------------------------\n")
	fmt.Printf("Wrote %v Tweets to %v\n in %v", total, args.OutputFile, urltmpl)
}